contracts/proxies/SafeProxy.sol

```diff
@@ -31,21 +31,25 @@ contract SafeProxy {
 
     /// @dev Fallback function forwards all transactions and returns all received return data.
     fallback() external payable {
-        // solhint-disable-next-line no-inline-assembly
+        /* solhint-disable no-inline-assembly */
+        /// @solidity memory-safe-assembly
         assembly {
             let _singleton := and(sload(0), 0xffffffffffffffffffffffffffffffffffffffff)
+            let ptr := mload(0x40)
             // 0xa619486e == keccak("masterCopy()"). The value is right padded to 32-bytes with 0s
             if eq(calldataload(0), 0xa619486e00000000000000000000000000000000000000000000000000000000) {
-                mstore(0, _singleton)
-                return(0, 0x20)
+                mstore(ptr, _singleton)
```

Comment:
I think using scratch space here is safe because we know that singleton is only 32 bytes and will not go beyond the 64 bytes scratch space.

From the docs:
>In particular, a memory-safe assembly block may only access the following memory ranges:

> Memory allocated by yourself using a mechanism like the allocate function described above.

> Memory allocated by Solidity, e.g. memory within the bounds of a memory array you reference.

> The scratch space between memory offset 0 and 64 mentioned above.

https://docs.soliditylang.org/en/v0.8.23/assembly.html#memory-safety

Reply:
But still, there's a higher-level question - do we want to make the peripheral contracts safe? It is safe to use as is (since we have the optimizer disabled), and with the optimizer enabled, it will not perform the optimizations unless the block is explicitly marked as memory-safe. It also compiles just fine afaik.

Reply:
We should use the scratch space here IMO (we explicitly only copy 32 bytes of return data which does fit into the scratch space and is considered memory safe according to the docs).

Reply:
Using scratch space inside the if block 👍🏾